pyomo/repn/plugins/gams_writer.py

```diff
@@ -0,0 +1,512 @@
+        put_results = io_options.pop("put_results", None)
+
+        if len(io_options):
+            raise ValueError(
+                "ProblemWriter_gams passed unrecognized io_options:\n\t" +
+                "\n\t".join("%s = %s"
+                            % (k,v) for k,v in io_options.iteritems()))
+
+        if solver is not None:
+            if solver.upper() not in valid_solvers:
+                raise ValueError("ProblemWriter_gams passed unrecognized "
+                                 "solver: %s" % solver)
+
+        if mtype is not None:
+            valid_mtypes = set([
+                'lp', 'qcp', 'nlp', 'dnlp', 'rmip', 'mip', 'rmiqcp', 'rminlp',
+                'miqcp', 'minlp', 'rmpec', 'mpec', 'mcp', 'cns', 'emp'])
+            if mtype.lower() not in valid_mtypes:
+                raise ValueError("ProblemWriter_gams passed unrecognized "
+                                 "model type: %s" % mtype)
+            if (solver is not None and
+                mtype.upper() not in valid_solvers[solver.upper()]):
+                raise ValueError("ProblemWriter_gams passed solver (%s) "
+                                 "unsuitable for given model type (%s)"
+                                 % (solver, mtype))
+
+        if output_filename is None:
+            output_filename = model.name + ".gms"
+
+        if symbolic_solver_labels and (labeler is not None):
+            raise ValueError("ProblemWriter_gams: Using both the "
+                             "'symbolic_solver_labels' and 'labeler' "
+                             "I/O options is forbidden")
+
+        if symbolic_solver_labels:
+            var_labeler = con_labeler = AlphaNumericTextLabeler()
+        elif labeler is None:
+            var_labeler = NumericLabeler('x')
+            con_labeler = NumericLabeler('c')
+        else:
+            var_labeler = con_labeler = labeler
+
+        var_list = []
+        symbolMap = SymbolMap()
+
+        def var_recorder(obj):
+            ans = var_labeler(obj)
+            var_list.append(ans)
+            return ans
+
+        def var_label(obj):
+            if obj.is_fixed():
+                return str(value(obj))
+            return symbolMap.getSymbol(obj, var_recorder)
+
+        # when sorting, there are a non-trivial number of
+        # temporary objects created. these all yield
+        # non-circular references, so disable GC - the
+        # overhead is non-trivial, and because references
+        # are non-circular, everything will be collected
+        # immediately anyway.
+        with PauseGC() as pgc:
+            try:
+                if isinstance(output_filename, string_types):
+                    output_file = open(output_filename, "w")
+                else:
+                    # Support passing of stream such as a StringIO
+                    # on which to write the model file
+                    output_file = output_filename
+                if isinstance(model, IBlockStorage):
```

Comment:
I finished implementing the more-or-less simple strategy of just passing labeler to to_string. I used grep to find all "def to_string" definitions in pyomo, added a labeler keyword to them, and passed labeler to all further calls of to_string within those definitions. ComponentData.to_string() and a new ICategorizedObject.to_string() are the only ones that apply the labeler if provided one. The GAMS solver now passes everything, but speaking of Baron, the Baron solver now errors on the following 4 tests:
```
ERROR: test_baron_bar_nonsymbolic_labels (pyomo.solvers.tests.checks.test_pickle.Test_LP_simple_kernel)
ERROR: test_baron_bar_symbolic_labels (pyomo.solvers.tests.checks.test_pickle.Test_LP_simple_kernel)
ERROR: test_baron_bar_nonsymbolic_labels (pyomo.solvers.tests.checks.test_pickle.Test_QP_simple_kernel)
ERROR: test_baron_bar_symbolic_labels (pyomo.solvers.tests.checks.test_pickle.Test_QP_simple_kernel)
```
I know very little about how Baron works, so I have no idea how to address these errors. The full error messages all look similar to this:
```
======================================================================
ERROR: test_baron_bar_nonsymbolic_labels (pyomo.solvers.tests.checks.test_pickle.Test_LP_simple_kernel)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gseastre/pyomo/src/pyomo/pyomo/solvers/tests/checks/test_pickle.py", line 54, in pickle_test
    load_solutions)
  File "/home/gseastre/pyomo/src/pyomo/pyomo/solvers/tests/models/base.py", line 113, in solve
    **io_options)
  File "/home/gseastre/pyomo/src/pyomo/pyomo/opt/base/solvers.py", line 623, in solve
    "Solver (%s) did not exit normally" % self.name)
ApplicationError: Solver (baron) did not exit normally
-------------------- >> begin captured stdout << ---------------------
ERROR: "[base]/pyomo/pyomo/opt/base/solvers.py", 616, solve
        Solver (baron) returned non-zero return code (3)
ERROR: "[base]/pyomo/pyomo/opt/base/solvers.py", 621, solve
        Solver log:
        ===========================================================================
         BARON version 17.4.1. Built: LNX-64 Sat Apr 1 11:24:13 EDT 2017 

         If you use this software, please cite:
         Tawarmalani, M. and N. V. Sahinidis, A polyhedral
         branch-and-cut approach to global optimization,
         Mathematical Programming, 103(2), 225-249, 2005.

         BARON is a product of The Optimization Firm, LLC. http://www.minlp.com/
         Parts of the BARON software were created at the
         University of Illinois at Urbana-Champaign.
        ===========================================================================
        ERROR: Illegal variable in line 19
        BARON: Syntax error.  Execution will stop

--------------------- >> end captured stdout << ----------------------
-------------------- >> begin captured logging << --------------------
pyomo.opt: ERROR: Solver (baron) returned non-zero return code (3)
pyomo.opt: ERROR: Solver log:
===========================================================================
 BARON version 17.4.1. Built: LNX-64 Sat Apr 1 11:24:13 EDT 2017 

 If you use this software, please cite:
 Tawarmalani, M. and N. V. Sahinidis, A polyhedral
 branch-and-cut approach to global optimization,
 Mathematical Programming, 103(2), 225-249, 2005.

 BARON is a product of The Optimization Firm, LLC. http://www.minlp.com/
 Parts of the BARON software were created at the
 University of Illinois at Urbana-Champaign.
===========================================================================
ERROR: Illegal variable in line 19
BARON: Syntax error.  Execution will stop 

--------------------- >> end captured logging << ---------------------
```